compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs

```diff
@@ -127,8 +127,8 @@ kindOfBuiltin = \case
 kindOf :: MonadGamma m => Type -> m Kind
 kindOf = \case
   TVar v -> lookupTypeVar v
-  TSyn{} -> error "TODO: kindOf, expand type synonym" -- TODO #3616
   TCon tcon -> kindOfDataType <$> inWorld (lookupDataType tcon)
+  TSynApp{} -> return KStar
```

Comment:
This check will occur in `expandTypeSynonyms`, which is called when an expresion is checked against the type. Is it really possible that this can't happen?

Reply:
You are assuming that there is such an expression. We should check that the type synonym definition is well-kinded even if there is no expression of that type.

Reply:
We do check the synonym definition - but that's unrelated to this issue.

This code is computing the kind of a type-expression, which take the form of a syn-application.

You are suggesting that LF may contain a type-expression for which we never check the type of an expression against.

Reply:
I can't imagine how this could happen. But I will repeat the well-formed check on the syn-application anyway.

Reply:
I’ve pushed a test case that demonstrates this problem. `checkDefTypeSyn` calls `checkType` which in turns just calls `kindOf` so that has to do the check.

Reply:
thanks. nice catch.
fixed now.